sdk/eventhub/Azure.Messaging.EventHubs/src/EventProcessorClient.cs

```diff
@@ -784,28 +797,54 @@ private async Task RunAsync(CancellationToken cancellationToken)
+                            .ForEach(ownership => partitionsOwnedByProcessorWithExactlyMaximumOwnedPartitionsCount.Add(ownership.PartitionId));
+                    }
+                    else if (activeOwnershipWithDistribution[key].Count > maximumOwnedPartitionsCount)
+                    {
+                        activeOwnershipWithDistribution[key]
+                            .ForEach(ownership => partitionsOwnedByProcessorWithGreaterThanMaximumOwnedPartitionsCount.Add(ownership.PartitionId));
+                    }
+                }
 
                 // Here's the important part.  If there are no processors that have exceeded the maximum owned partition count allowed, we may
                 // need to steal from the processors that have exactly the maximum amount.  If this instance is below the minimum count, then
                 // we have no choice as we need to enforce balancing.  Otherwise, leave it as it is because the distribution wouldn't change.
 
-                if (!stealablePartitions.Any() && ownedPartitionsCount < minimumOwnedPartitionsCount)
+                if (!partitionsOwnedByProcessorWithGreaterThanMaximumOwnedPartitionsCount.Any() && ownedPartitionsCount < minimumOwnedPartitionsCount)
                 {
-                    stealablePartitions = activeOwnership
-                        .Where(ownership => partitionDistribution[ownership.OwnerIdentifier] == maximumOwnedPartitionsCount)
-                        .Select(ownership => ownership.PartitionId);
-                }
+                    // If any stealable partitions were found, randomly pick one of them to claim.
+                    var index = RandomNumberGenerator.Value.Next(partitionsOwnedByProcessorWithExactlyMaximumOwnedPartitionsCount.Count());
 
-                // If any stealable partitions were found, randomly pick one of them to claim.
-
-                if (stealablePartitions.Any())
+                    return await ClaimOwnershipAsync(
+                        partitionsOwnedByProcessorWithExactlyMaximumOwnedPartitionsCount.ElementAt(index),
+                        completeOwnershipEnumerable)
+                        .ConfigureAwait(false);
+                }
+                else if (partitionsOwnedByProcessorWithGreaterThanMaximumOwnedPartitionsCount.Any())
                 {
-                    var index = RandomNumberGenerator.Value.Next(stealablePartitions.Count());
+                    // If any stealable partitions were found, randomly pick one of them to claim.
+                    var index = RandomNumberGenerator.Value.Next(partitionsOwnedByProcessorWithGreaterThanMaximumOwnedPartitionsCount.Count());
 
-                    return await ClaimOwnershipAsync(stealablePartitions.ElementAt(index), completeOwnershipEnumerable).ConfigureAwait(false);
+                    return await ClaimOwnershipAsync(
+                        partitionsOwnedByProcessorWithGreaterThanMaximumOwnedPartitionsCount.ElementAt(index),
```

Comment:
nit: I believe that you can use the indexer directly here rather than `ElementAt`